etc/statistic_summary_single_column.c

```diff
@@ -0,0 +1,316 @@
+
+/*
+   Simplified sqrt within 1 ulp, no rounding; good enough to compute the SD
+
+   From a copy of a drafted paper by Prof W. Kahan and K.C. Ng, written in May, 1986.
+   That copy can be found in the file https://www.netlib.org/fdlibm/e_sqrt.c
+*/
+static double s_sqrt(double d)
+{
+   uint32_t high;
+
+   uint32_t k, y0;
+   double y;
+   uint32_t T1[32] = {
+      0,      1024,   3062,   5746,   9193,   13348,  18162,  23592,
+      29598,  36145,  43202,  50740,  58733,  67158,  75992,  85215,
+      83599,  71378,  60428,  50647,  41945,  34246,  27478,  21581,
+      16499,  12183,  8588,   5674,   3403,   1742,   661,    130
+   };
+
+   union {
+      double d;
+      uint64_t u;
+   } guts;
+
+   guts.d = d;
+
+   if (((guts.u >= 0x7FF0000000000001ULL) && (guts.u <= 0x7FF7FFFFFFFFFFFFULL))
+       || ((guts.u >= 0xFFF0000000000001ULL) && (guts.u <= 0xFFF7FFFFFFFFFFFFULL))
+       || ((guts.u >= 0x7FF8000000000001ULL) && (guts.u <= 0x7FFFFFFFFFFFFFFFULL))
+       || ((guts.u >= 0xFFF8000000000001ULL) && (guts.u <= 0xFFFFFFFFFFFFFFFFULL))
+       || (guts.u == 0x7FF0000000000000ULL)
+       || (guts.u == 0x0000000000000000ULL)
+       || (guts.u == 0x8000000000000000ULL)) {
+      return d;
+   }
+
+   if (d < 0.0) {
+      guts.u = 0xFFF8000000000000ULL;
+      return guts.d;
+   }
+
+
+   high = guts.u >> 32;
+
+   k = (high >> 1) + 0x1ff80000UL;
+   y0 = k - T1[31 & (k >> 15)];
+
+   guts.u = 0uLL;
+   guts.u = y0;
+   guts.u <<= 32;
+
+   y = guts.d;
+
+   y = (y + d/y) / 2.0;
+   y = (y + d/y) / 2.0;
+   y = y - (y - d/y) / 2.0;
+
+   return y;
+}
+
+static double s_sd(long *array, int length, double mean)
+{
+   double sum = 0.0;
+   int i;
+   for (i = 0; i < length; i++) {
+      sum += (array[i] - mean) * (array[i] - mean);
+   }
+   return s_sqrt(sum/(length-1));
+}
+
+static double s_sd_kurt(long *array, int length, double mean)
+{
+   double sum = 0.0;
+   int i;
+   for (i = 0; i < length; i++) {
+      sum += (array[i] - mean) * (array[i] - mean);
+   }
+   return s_sqrt(sum/length);
+}
+
+static double s_kurtosis(long *array, int length, double mean)
+{
+   double sum = 0.0, quad = 0.0;
+   int i;
+   for (i = 0; i < length; i++) {
+      quad = (array[i] - mean);
+      quad = quad * quad;
+      quad = quad * quad;
+      sum += quad;
+   }
+   quad = s_sd_kurt(array, length, mean);
+   quad = quad * quad;
+   quad = quad * quad;
+   return ((sum/(length)) / quad);
+}
+
+static double s_skewness(long *array, int length, double mean)
+{
+   double sum = 0.0, cube = 0.0;
+   int i;
+   for (i = 0; i < length; i++) {
+      cube = (array[i] - mean);
+      sum += cube * cube * cube;
+   }
+   cube = s_sd_kurt(array, length, mean);
+   cube = cube * cube * cube;
+   return ((sum/(length)) / cube);
+}
+
+/*
+   Algorithm found in Section 2.3.4 "Finding the Median" in
+   Wirth, Niklaus. "Algorithms + data structures = programs."
+   No. 04; QA76. 6, W56.. 1976.
+*/
+static long k_smallest(long *array, int length, int k)
```

Comment:
Why not qsort + lookup? Is this performance critical?

Reply:
No qsort in LTM (grep for "unwanted" in `helper.pl`). It's more of a rule for LTM itself but I had no reason to break with it here.

Reply:
This rule should be broken. The same argument as above applies.

Reply:
Yes, but as above: it is automated now and part of the buildprocess, so the same restrictions as for LTM itself apply.

Reply:
I think the median is acceptable since it's a better algorithm than qsort.